compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2240,6 +2240,174 @@ OMR::Power::TreeEvaluator::lbitpermuteEvaluator(TR::Node *node, TR::CodeGenerato
+
+   if (type != TR::Int64)
+      {
+      TR::InstOpCode::Mnemonic loadOpcode;
+      uint64_t zeroExtendMask;
+      uint32_t length;
+      switch (type)
+         {
+         case TR::Int32:
+            loadOpcode = TR::InstOpCode::lwz;
+            zeroExtendMask = CONSTANT64(0x00000000ffffffff);
+            length = 4;
+            break;
+         case TR::Int16:
+            loadOpcode = TR::InstOpCode::lhz;
+            zeroExtendMask = CONSTANT64(0x000000000000ffff);
+            length = 2;
+            break;
+         case TR::Int8:
+            loadOpcode = TR::InstOpCode::lbz;
+            zeroExtendMask = CONSTANT64(0x00000000000000ff);
+            length = 1;
+            break;
+         default:
+            TR_ASSERT_FATAL(false, "Unrecognized compressbits type %s\n", type.toString());
+            break;
+         }
+
+      // Only one of src or mask needs to be zero-extended
+      // First check if either can be zero-extended at compile time
+      if (maskNode->getOpCode().isLoadConst())
+         {
+         int64_t mask = maskNode->getUnsignedLongInt() & zeroExtendMask;
+         maskReg = cg->allocateRegister();
+         loadConstant(cg, maskNode, mask, maskReg);
+         resReg = maskReg;
+         }
+      else if (srcNode->getOpCode().isLoadConst())
+         {
+         int64_t src = srcNode->getUnsignedLongInt() & zeroExtendMask;
+         srcReg = cg->allocateRegister();
+         loadConstant(cg, srcNode, src, srcReg);
+         resReg = srcReg;
+         }
+      // Otherwise check if either is in memory and can therefore be loaded with a zero-extending opcode
+      else if (maskNode->getReferenceCount() == 1 && maskNode->getOpCode().isMemoryReference() && maskNode->getRegister() == NULL)
+         {
+         maskReg = cg->allocateRegister();
+         TR::LoadStoreHandler::generateLoadNodeSequence(cg, maskReg, maskNode, loadOpcode, length);
+         resReg = maskReg;
+         }
+      else if (srcNode->getReferenceCount() == 1 && srcNode->getOpCode().isMemoryReference() && srcNode->getRegister() == NULL)
+         {
+         srcReg = cg->allocateRegister();
+         TR::LoadStoreHandler::generateLoadNodeSequence(cg, srcReg, srcNode, loadOpcode, length);
+         resReg = srcReg;
+         }
+      // If no zero-extension shortcut applied, default to zero-extending the mask
+      if (resReg == NULL)
+         {
+         maskReg = cg->gprClobberEvaluate(maskNode);
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicl, node, maskReg, maskReg, 0, zeroExtendMask);
+         resReg = maskReg;
+         }
+      // Evaluate the missing operand
+      if (maskReg == NULL)
+         {
+         maskReg = cg->evaluate(maskNode);
+         }
+      else
+         {
+         srcReg = cg->evaluate(srcNode);
+         }
+      }
+   else
+      {
+      srcReg = cg->evaluate(srcNode);
+      maskReg = cg->gprClobberEvaluate(maskNode);
+      resReg = maskReg;
+      }
```

Comment:
are these methods only existing in java versions later than java8? otherwise, you need to consider for 32bit JVMs.

Reply:
This is an important point.  The methods only exist in Java 19 and later.  However, as these are fully-fledged operators in the IL, It's conceivable that they will be used in other contexts, so 32-bit support would still be needed.